site/app/libraries/NotebookUtils.php

```diff
@@ -5,90 +5,254 @@
 namespace app\libraries;
 
 class NotebookUtils {
+    private const TEXT_LIMIT = 1024 * 5; // 5KB limit for output text
+    private const LINE_LIMIT = 100; // 100 lines limit for text output
+    private const IMG_SIZE_LIMIT = 1024 * 1024 * 2; // 2MB limit for images
+    private const NOTEBOOK_SIZE_LIMIT = 1024 * 1024 * 10; // 10MB limit for rendering
+    // Note: SVG files are not supported due to XSS risks
+    private const MIME_TYPES = [
+        'image/png',
+        'image/jpeg',
+        'image/gif',
+        'image/bmp',
+        'text/plain', // Fallback to text/plain if it is available.
+    ];
+
     /**
      * Accepts a path to a .ipynb file and returns an array in the Submitty notebook format.
      *
      * @return array<int,array<string,mixed>>
      */
     public static function jupyterToSubmittyNotebook(string $filepath): array {
+        // Check the total file size before doing any processing
+        if (filesize($filepath) > self::NOTEBOOK_SIZE_LIMIT) {
+            // Return a single markdown cell with an error message
+            return [
+                [
+                    'type' => 'markdown',
+                    'markdown_data' => '### Notebook Not Rendered' . PHP_EOL .
+                    'The notebook file is too large to be displayed (over ' . (self::NOTEBOOK_SIZE_LIMIT / (1024 * 1024)) . ' MB).' . PHP_EOL .
+                    'Please download the notebook to view its contents.'
+                ]
+            ];
+        }
+
         $filedata = FileUtils::readJsonFile($filepath);
         if ($filedata === false) {
             $filedata = [];
         }
 
         $cells = [];
-
+        $skipped_content = 0;
+        $skipped_output = 0;
```

Comment:
Optional change:

I found these names confusing.  My first guess was that they referred to either the actual content being skipped, or cells being skipped, when in fact they refer to the _number_ of cells being skipped.  Perhaps rename?